### PR TITLE
[Test] Batch Invariant: Rename and organize tests

### PR DESCRIPTION
**Pull request by yewentao256**
## Purpose

Refactor the unit test folder to reuse the utils.

Add online test

## Test

Online test example

- `export VLLM_ATTENTION_BACKEND="FLASHINFER_MLA"`
- `export VLLM_BATCH_INVARIANT=1`
- `vllm serve deepseek-ai/DeepSeek-R1 -tp 8  --enable-expert-parallel --port 9256`
- `pytest tests/v1/determinism/test_online_batch_invariance.py`

```bash
(wentao) wentao@dgxB200-09:~/vllm-source$ pytest tests/v1/generation/test_online_batch_invariance.py
============================================ test session starts =============================================
platform linux -- Python 3.12.3, pytest-8.4.1, pluggy-1.6.0
rootdir: /home/wentao/vllm-source
configfile: pyproject.toml
plugins: anyio-4.11.0
collected 1 item                                                                                             

tests/v1/generation/test_online_batch_invariance.py .                                                  [100%]
================================= 1 passed, 2 warnings in 165.79s (0:02:45) ==================================
```

### CODE REVIEW

File: vllm/v1/attention/backends/mla/flashinfer_mla.py [hunk 1]
```diff
@@ -127,19 +130,43 @@ def _forward_decode(
...
+                    kv_lora_rank=self.kv_lora_rank,
+                    qk_rope_head_dim=self.qk_rope_head_dim,
+                    block_tables=bt_i,
+                    seq_lens=sl_i,
+                    max_seq_len=attn_metadata.max_seq_len,
+                    bmm1_scale=self.bmm1_scale,
+                    bmm2_scale=self.bmm2_scale,
+                )
+                outs.append(oi)
+            o = torch.cat(outs, dim=0)
```
[the remark this answers is not in the document]
Comment (yewentao256) [reply vllm/v1/attention/backends/mla/flashinfer_mla.py]: `num = q.shape[0]` shouldn't be equal to 0

File: vllm/v1/attention/backends/mla/flashinfer_mla.py [hunk 2]
```diff
@@ -127,19 +130,43 @@ def _forward_decode(
...
-            kv_lora_rank=self.kv_lora_rank,
-            qk_rope_head_dim=self.qk_rope_head_dim,
-            block_tables=attn_metadata.decode.block_table,
-            seq_lens=attn_metadata.decode.seq_lens,
-            max_seq_len=attn_metadata.max_seq_len,
-            bmm1_scale=self.bmm1_scale,
-            bmm2_scale=self.bmm2_scale,
-        )
+        if vllm_is_batch_invariant():
+            # execute per-request to eliminate batch-shape-dependent kernel paths.
```
Comment (bwasti) [inline vllm/v1/attention/backends/mla/flashinfer_mla.py]: ah nice.  is flash-infer going to officially support this path too? (like they do with non-MLA)

cc @pavanimajety

Comment (yewentao256) [reply vllm/v1/attention/backends/mla/flashinfer_mla.py]: I think we can land this PR first, and if it is supported officially later, we can have a following up PR to optimize.
@mgoin CC

File: tests/v1/generation/conftest.py [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+def enable_batch_invariant_mode():
+    """Automatically enable batch invariant kernel overrides for all tests."""
+    old_value = os.environ.get("VLLM_BATCH_INVARIANT")
+    os.environ["VLLM_BATCH_INVARIANT"] = "1"
+    yield
+    # restore original value after test
+    if old_value is None:
+        os.environ.pop("VLLM_BATCH_INVARIANT", None)
+    else:
+        os.environ["VLLM_BATCH_INVARIANT"] = old_value
```
Comment (mgoin) [inline tests/v1/generation/conftest.py]: Can we use monkeypatch instead? Also, why is this folder called generation if it is only for batch invariance? We should rename to relate to determinism

File: vllm/entrypoints/openai/serving_completion.py [hunk 4]
```diff
@@ -206,13 +206,27 @@ async def create_completion(
...
+                    data_parallel_rank = None
+                    from vllm.model_executor.layers.batch_invariant import (
+                        vllm_is_batch_invariant,
+                    )
+
+                    if vllm_is_batch_invariant():
+                        config = self.engine_client.vllm_config
+                        dp_size = config.parallel_config.data_parallel_size
+                        if dp_size and dp_size > 1:
+                            data_parallel_rank = 0
```
Comment (mgoin) [inline vllm/entrypoints/openai/serving_completion.py]: Why is this needed? If each dp rank itself is batch invariant, then it shouldn't matter what rank gets the request

File: vllm/v1/attention/backends/mla/flashinfer_mla.py [hunk 5]
```diff
@@ -127,19 +130,43 @@ def _forward_decode(
...
+            num = q.shape[0]
+            outs = []
+            for i in range(num):
+                qi = q[i : i + 1]
+                bt_i = attn_metadata.decode.block_table[i : i + 1]
+                sl_i = attn_metadata.decode.seq_lens[i : i + 1]
+                oi = trtllm_batch_decode_with_kv_cache_mla(
```
Comment (mgoin) [inline vllm/v1/attention/backends/mla/flashinfer_mla.py]: I don't have a clear idea for what features we need to make batch invariant, but this naive loop seems like a bad pattern. This is clearly a massive loss of performance so I think we should instead look to work with the kernel itself to make it stable

**Review by mgoin — CHANGES_REQUESTED** [review]
I feel like these patches to hardcode DP rank 0 and executing mla one token at a time are not a good way to achieve batch invariance. This seems like a slippery slope where we could just do hacks like this for all kernels, which I think we clearly don't want

File: tests/v1/generation/conftest.py (hunk 3 above)
Comment (yewentao256) [reply tests/v1/generation/conftest.py]: monkeypatch is used in https://github.com/vllm-project/vllm/pull/27478.
After that is landed, we can further more rename to determinism.

File: vllm/entrypoints/openai/serving_completion.py (hunk 4 above)
Comment (yewentao256) [reply vllm/entrypoints/openai/serving_completion.py]: Yes—this is necessary. Batch invariance is per DP rank; with DP>1 the same request can land on different ranks between bs=1 and bs=N (eg, 15 prompts to rank 0 and 17 goes to rank1, is totally different with all to 0), and ranks can legitimately differ, causing the bitwise test to fail.

File: vllm/v1/attention/backends/mla/flashinfer_mla.py (hunk 5 above)
Comment (yewentao256) [reply vllm/v1/attention/backends/mla/flashinfer_mla.py]: Yes I agree with you, added a TODO and we should fix this in following up PR. (We may need to update upstream Flashinfer so it may take a while)

File: tests/v1/generation/conftest.py (hunk 3 above)
Comment (yewentao256) [reply tests/v1/generation/conftest.py]: Done

File: vllm/entrypoints/openai/serving_completion.py (hunk 4 above)
Comment (mgoin) [reply vllm/entrypoints/openai/serving_completion.py]: You say this "ranks can legitimately differ" but I don't understand why this would be the case. In DP the layers will be duplicated to be the same size between ranks, thus calling the same kernels

Comment (yewentao256) [reply vllm/entrypoints/openai/serving_completion.py]: I meant we have batch requests, one dp rank gets 32 as a batch, and one dp gets 15 and another gets 17, it is not the same. The batch invariant requires same logits for the same rank instead of different ranks. We haven't supported "parallism invariant" yet.

File: tests/v1/determinism/test_online_batch_invariance.py
Comment (mgoin) [inline tests/v1/determinism/test_online_batch_invariance.py:1]: This seems like it implements a custom server wrapper and client for no clear reason. Can you use the usual RemoteOpenAIServer interface we use in api server tests?
https://github.com/vllm-project/vllm/blob/f08eab2acc17da9e86d20673bd801659ca912749/tests/entrypoints/openai/test_chat_logit_bias_validation.py#L24-L41

**Review by mgoin — CHANGES_REQUESTED** [review]

File: tests/v1/determinism/test_online_batch_invariance.py
Comment (mgoin) [reply tests/v1/determinism/test_online_batch_invariance.py:1]: Also is there a reason why DeepSeek-R1 is needed for this test? I feel like we could just as well run a deepseek lite or qwen3 fp8 to test similar features on 2xH100 or 2xB200, and actually enable this test in CI

Comment (yewentao256) [reply tests/v1/determinism/test_online_batch_invariance.py:1]: Fixed, thanks! Could you take another look? @mgoin

**Review by mgoin — APPROVED** [review]
LGTM, thanks!
